crates/core/src/polynomial/arith_circuit.rs

```diff
@@ -159,22 +216,28 @@ impl<F: TowerField> DeserializeBytes for ArithCircuitPoly<F> {
 	where
 		Self: Sized,
 	{
-		let (expr, n_vars) = <(ArithExpr<F>, usize)>::deserialize(read_buf, mode)?;
-		Self::with_n_vars(n_vars, expr).map_err(|_| SerializationError::InvalidConstruction {
-			name: "ArithCircuitPoly",
+		let (circuit, n_vars) = <(ArithCircuit<F>, usize)>::deserialize(read_buf, mode)?;
+		Self::with_n_vars_circuit(n_vars, circuit).map_err(|_| {
+			SerializationError::InvalidConstruction {
+				name: "ArithCircuitPoly",
+			}
 		})
 	}
 }
 
 impl<F: TowerField> ArithCircuitPoly<F> {
-	pub fn new(expr: ArithExpr<F>) -> Self {
-		let degree = expr.degree();
-		let n_vars = expr.n_vars();
-		let tower_level = expr.binary_tower_level();
-		let (exprs, retval) = circuit_steps_for_expr(&expr);
+	pub fn new(expr: &ArithExpr<F>) -> Self {
```

Comment:
There are a lot of lines in the diff because of this change. No good reason to change it back at this point, but it should be fine to have kept it as it was -- cloning Arcs is cheap.